contracts/LimitOrderSwap.sol

```diff
@@ -0,0 +1,207 @@
+    // whether an order is canceled or not
+    mapping(bytes32 => bool) public orderHashToCanceled;
+
+    constructor(
+        address _owner,
+        address _uniswapPermit2,
+        address _allowanceTarget,
+        IWETH _weth,
+        address payable _feeCollector
+    ) Ownable(_owner) TokenCollector(_uniswapPermit2, _allowanceTarget) {
+        weth = _weth;
+        feeCollector = _feeCollector;
+    }
+
+    receive() external payable {}
+
+    /// @notice Only owner can call
+    /// @param _newFeeCollector The new address of fee collector
+    function setFeeCollector(address payable _newFeeCollector) external onlyOwner {
+        if (_newFeeCollector == address(0)) revert ZeroAddress();
+        feeCollector = _newFeeCollector;
+
+        emit SetFeeCollector(_newFeeCollector);
+    }
+
+    /// @inheritdoc ILimitOrderSwap
+    function fillLimitOrder(
+        LimitOrder calldata order,
+        bytes calldata makerSignature,
+        TakerParams calldata takerParams
+    ) external payable override {
+        _fillLimitOrder(order, makerSignature, takerParams, false);
+    }
+
+    /// @inheritdoc ILimitOrderSwap
+    function fillLimitOrderFullOrKill(
+        LimitOrder calldata order,
+        bytes calldata makerSignature,
+        TakerParams calldata takerParams
+    ) external payable override {
+        _fillLimitOrder(order, makerSignature, takerParams, true);
+    }
+
+    /// @inheritdoc ILimitOrderSwap
+    function cancelOrder(LimitOrder calldata order) external override {
+        if (order.expiry <= uint64(block.timestamp)) revert ExpiredOrder();
+        if (msg.sender != order.maker) revert NotOrderMaker();
+        bytes32 orderHash = getLimitOrderHash(order);
+        uint256 orderFilledAmount = orderHashToMakerTokenFilledAmount[orderHash];
+        if (orderFilledAmount >= order.makerTokenAmount) revert FilledOrder();
+        if (orderHashToCanceled[orderHash]) revert CanceledOrder();
+
+        // Set canceled state to storage
+        orderHashToCanceled[orderHash] = true;
+        emit OrderCanceled(orderHash, order.maker);
+    }
+
+    function _fillLimitOrder(
+        LimitOrder calldata order,
+        bytes calldata makerSignature,
+        TakerParams calldata takerParams,
+        bool fullOrKill
+    ) private {
+        (bytes32 orderHash, uint256 takerSpendingAmount, uint256 makerSpendingAmount) = _validateOrderAndQuote(
+            order,
+            makerSignature,
+            takerParams.takerTokenAmount,
+            takerParams.makerTokenAmount,
+            fullOrKill
+        );
+
+        // maker -> taker
+        _collect(order.makerToken, order.maker, address(this), makerSpendingAmount, order.makerTokenPermit);
+        uint256 fee = (makerSpendingAmount * order.feeFactor) / Constant.BPS_MAX;
+        order.makerToken.transferTo(payable(takerParams.recipient), makerSpendingAmount - fee);
+        // collect fee if present
+        order.makerToken.transferTo(feeCollector, fee);
```

Comment:
nit: Not sure if checking fee > 0 before transferring could save some gas for no fee case? But if we always have fee, then this might not worth to add. 😆